streams/src/main/java/org/apache/kafka/streams/query/VersionedKeyQuery.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.query;
+
+import java.time.Instant;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.kafka.common.annotation.InterfaceStability.Evolving;
+import org.apache.kafka.streams.state.VersionedRecord;
+
+/**
+ * Interactive query for retrieving a single record from a versioned state store based on its key and timestamp.
+ */
+@Evolving
+public final class VersionedKeyQuery<K, V> implements Query<VersionedRecord<V>> {
+
+    private final K key;
+    private final Optional<Instant> asOfTimestamp;
+
+    private VersionedKeyQuery(final K key, final Optional<Instant> asOfTimestamp) {
+        this.key = Objects.requireNonNull(key);
+        this.asOfTimestamp = asOfTimestamp;
+    }
+
+    /**
+     * Creates a query that will retrieve the record from a versioned state store identified by {@code key} if it exists
+     * (or {@code null} otherwise).
+     * @param key The key to retrieve
+     * @param <K> The type of the key
+     * @param <V> The type of the value that will be retrieved
+     * @throws NullPointerException if @param key is null
+     */
+    public static <K, V> VersionedKeyQuery<K, V> withKey(final K key) {
+        return new VersionedKeyQuery<>(key, Optional.empty());
+    }
+
+    /**
+     * Specifies the timestamp for the key query. The key query returns the record version for the specified timestamp.
+     * (To be more precise: The key query returns the record with the greatest timestamp <= asOfTimestamp)
+     * if @param asOfTimestamp is null, it will be considered as Optional.empty()
```

Comment:
> Did we discuss this in the KIP (cannot remember) -- by gut feeling would be that `null` should not be allowed (because not necessary, as "latest" is the default anyway)?

We discussed in the other kips that if the time bound is specified as null, the query should not return NPE. But you are right about this query type; the default is always the latest. Therefore, we can forbid `null` here.